### PR TITLE
fix(compiler): avoid converting &nbps; to spaces

### LINKED ISSUE #10485: With `whitespace: 'condense'`, elements contain only &nbsp; and spaces are condensed into empty tags
### Version
4.0.0-rc.1

### Reproduction link
[https://github.com/tobyzerner/vue-cli-nbsp-bug](https://github.com/tobyzerner/vue-cli-nbsp-bug)

### Environment info
```
Environment Info:
  System:
    OS: macOS 10.14.6
    CPU: (4) x64 Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
  Binaries:
    Node: 10.15.3 - /usr/local/bin/node
    Yarn: 1.17.3 - ~/.npm-global/bin/yarn
    npm: 6.11.3 - ~/.npm-global/bin/npm
  Browsers:
    Chrome: 76.0.3809.132
    Firefox: 68.0.2
    Safari: 12.1.2
  npmPackages:
    @vue/babel-helper-vue-jsx-merge-props:  1.0.0
    @vue/babel-plugin-transform-vue-jsx:  1.0.0
    @vue/babel-preset-app:  4.0.0-rc.1
    @vue/babel-preset-jsx:  1.1.0
    @vue/babel-sugar-functional-vue:  1.0.0
    @vue/babel-sugar-inject-h:  1.0.0
    @vue/babel-sugar-v-model:  1.0.0
    @vue/babel-sugar-v-on:  1.1.0
    @vue/cli-overlay:  4.0.0-rc.1
    @vue/cli-plugin-babel: ^4.0.0-rc.1 => 4.0.0-rc.1
    @vue/cli-plugin-eslint: ^4.0.0-rc.1 => 4.0.0-rc.1
    @vue/cli-service: ^4.0.0-rc.1 => 4.0.0-rc.1
    @vue/cli-shared-utils:  4.0.0-rc.1
    @vue/component-compiler-utils:  3.0.0
    @vue/preload-webpack-plugin:  1.1.1
    @vue/web-component-wrapper:  1.2.0
    eslint-plugin-vue: ^5.0.0 => 5.2.3
    vue: ^2.6.10 => 2.6.10
    vue-eslint-parser:  5.0.0
    vue-hot-reload-api:  2.3.3
    vue-loader:  15.7.1
    vue-style-loader:  4.1.2
    vue-template-compiler: ^2.6.10 => 2.6.10
    vue-template-es2015-compiler:  1.9.1
  npmGlobalPackages:
    @vue/cli: 4.0.0-rc.1
```

### Steps to reproduce
Clone the repo, `npm install && npm run serve`

### What is expected?
The template contains a span containing a `&nbsp;` character. This span has a red background. Thus you should see a small red box on the page.

### What is actually happening?
The `&nbsp;` in the template is being ignored and converted into a normal space, which means the span element is not displayed.

---
I cannot replicate this bug in a plain Vue jsfiddle, nor a project using vue-loader without vue-cli. Thus I think something in vue-cli may be the culprit rather than vue-template-compiler or Vue itself.

### LINKED ISSUE #11059: With `whitespace:condense` all &nbsps;, even meaningful, are converted to spaces
### Version
2.6.11

### Reproduction link
[https://gist.github.com/aimozg/c073eea5a55062e3dcd8f887a8567807](https://gist.github.com/aimozg/c073eea5a55062e3dcd8f887a8567807)

### Steps to reproduce
Run example.js in Node.

Alternative steps/Detailed explanation:
1. Setup Vue template compiler with `whitespace: 'condense'`. (it is default for vue-cli).
2. Render template with `&nbsp;` entity, e.g. `<p>a&nbsp;b</p>`
3. Inspect template for actual non-breaking space character

### What is expected?
`&nbsp;` entity rendered as non-breaking space character; example.js output should contain
```
---- whitespace: condense
// code
NBSP FOUND
```

### What is actually happening?
`&nbsp;` entity is rendered as plain space, example.js output contains
```
---- whitespace: condense
// code
NBSP NOT FOUND
```

---
The whitespace is stripped by this compiler code:

```js
if (!inPre && whitespaceOption === 'condense') {
  // condense consecutive whitespaces into single space
  text = text.replace(whitespaceRE, ' ')
}
```

and `"&nbsp;"`s are replaced by `"\xA0"`s with entity decoder earlier.

Related issues is https://github.com/vuejs/vue/issues/10485

**Suggestion**:

Narrower whitespaceRE regexp, like `/[ \t\r\n]+/g`.

### PR DESCRIPTION
**Pull request by despreston**
The regex for whitespace was too strict and was causing $nbps; chars to
disappear from templates when `whitespace: 'condense'` is set. Changing the rule
to avoid converting non-breaking white space chars into regular spaces.

Fixes #10485
Fixes #11059

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

### CODE REVIEW

File: src/compiler/parser/index.js [hunk 1]
```diff
@@ -38,7 +38,7 @@ const modifierRE = /\.[^.\]]+(?=[^\]]*$)/g
 const slotRE = /^v-slot(:|$)|^#/
 
 const lineBreakRE = /[\r\n]/
-const whitespaceRE = /\s+/g
+const whitespaceRE = /[ \t\r\n]+/g
```
Comment (Justineo) [inline src/compiler/parser/index.js]: Maybe we should use *space characters* defined in HTML spec [here](https://www.w3.org/TR/html52/infrastructure.html#space-characters)?

Comment (despreston) [reply src/compiler/parser/index.js]: Yeah, thats a good idea. The only possible char missing from the regex at this point would be `U+000C FORM FEED (FF)`

**Comment by screetBloom** [comment]
What is the expected version number of this pr?

@Justineo @yyx990803

**Comment by screetBloom** [comment]
> What is the expected version number of this pr, And when will the patch version be released?
> 

Ask again~ , Can someone answer my question
@posva  @yyx990803  @Justineo

**Comment by Justineo** [comment]
As planned earlier, we'll come back for Vue 2.x after Vue 3 is officially released.

**Comment by despreston** [comment]
hey @Justineo just want to check in. Any updates planned for this one?

**Comment by despreston** [comment]
@posva We're not really getting much of a response regarding a timeline for this one. Any idea when this can ship?

**Comment by Matt-Mass** [comment]
I also have the problem. is there any advancement on this subject ?

**Review by posva — APPROVED** [review]

**Pull request merged**